### PR TITLE
arch: refractor eventstream into async

### PR DESCRIPTION
**Pull request by xingyaoww**
**What is the problem that this fixes or functionality that this introduces? Does it fix any open issues?**

Context: https://opendevin.slack.com/archives/C06QKSD9UBA/p1720670497261339

Refractor the `Runtime` class to `async`-style so that it supports async `.close` required by the `RuntimeClient` in order to close the `AsyncSession`.

Dependency: need to merge https://github.com/OpenDevin/OpenDevin/pull/2900 first.

**Give a brief summary of what the PR does, explaining any non-trivial design decisions**

- Add async `initialize` to `Runtime` class
- Make `Runtime` a more generic parent class since `EventStreamRuntime` inherits from it and has a much different implementation than the current default `ServerRuntime`.
  - Move most `Runtime` logic to its child class `ServerRuntime` since they are more tailored for `ServerRuntime`
- Mark two of the `Runtime` method as deprecated (once `EventStreamRuntime` is ready).

**Other references**

### CODE REVIEW

**Review by yufansong — APPROVED** [review]
LGTM on my side.

File: opendevin/runtime/runtime.py [hunk 1]
```diff
@@ -58,50 +38,53 @@ class Runtime:
...
-        else:
-            self.sandbox = sandbox
-            self._is_external_sandbox = True
-        self.browser: BrowserEnv | None = None
-        self.file_store = InMemoryFileStore()
         self.event_stream = event_stream
         self.event_stream.subscribe(EventStreamSubscriber.RUNTIME, self.on_event)
+        atexit.register(self.close_sync)
+
+    async def initialize(self) -> None:
```
Comment (tobitege) [inline opendevin/runtime/runtime.py:47]: Like with the other class, maybe rename to init_async (or ainit), just an idea.

**Review by tobitege — APPROVED** [review]
Only some notes/ideas, overall looks good.
Don't have the time to actually run this branch myself, though.

File: opendevin/runtime/runtime.py (hunk 1 above)
Comment (xingyaoww) [reply opendevin/runtime/runtime.py]: renamed to `ainit` :D

File: opendevin/runtime/client/runtime.py [hunk 2]
```diff
@@ -121,9 +115,8 @@ def _init_container(
                     'PYTHONUNBUFFERED=1 poetry run '
                     f'python -u -m opendevin.runtime.client.client {self._port} '
                     f'--working-dir {sandbox_workspace_dir} '
-                    f'--plugins {plugin_names} '
+                    f'--plugins {plugin_names}'
```
[the remark this answers is not in the document]
Comment (xingyaoww) [reply opendevin/runtime/client/runtime.py:118]: I plan to tests this Runtime extensively by writing a bunch of tests for it as the next step :D - We will probably try tweaking these argument as empty and see it it break things

**Pull request merged**